File: sdk/desktop/macos/Sources/KsealDesktop/Security/HardwareKeyStore.swift
import Foundation
#if canImport(Security)
import Security
#endif
#if canImport(CryptoKit)
import CryptoKit
#endif

/// Binds at-rest key material to a hardware-backed key where the platform offers
/// one, with a clean software fallback. The request-proof HMAC key is sealed
/// before it is persisted, so the on-disk secret cannot be lifted and replayed
/// from another device — yet the **request-proof byte layout is unchanged**: the
/// core still computes `HMAC(proofKey, …)`; only how `proofKey` is protected at
/// rest changes.
///
/// Production binds to the macOS Keychain (Secure-Enclave-wrapped on Apple
/// silicon). This is the **external secure-element mock boundary**: tests inject
/// a fake store so the proof-binding logic is exercised deterministically on any
/// host.
public protocol HardwareKeyStore {
    /// Whether sealed blobs are bound to a hardware-backed key (vs. a software
    /// fallback). Surfaced so a policy can require hardware binding and fail
    /// closed when it is unavailable.
    var isHardwareBacked: Bool { get }

    /// Wraps `plaintext` into an opaque blob safe to persist. Throws if the
    /// hardware element rejects the operation.
    func seal(_ plaintext: Data) throws -> Data

    /// Unwraps a blob previously produced by `seal`. Throws if the blob was not
    /// produced by this store / device (tamper or wrong-device).
    func unseal(_ sealed: Data) throws -> Data
}

/// Raised when a hardware key operation cannot complete.
public struct HardwareKeyStoreError: Error, CustomStringConvertible {
    public let message: String
    public var description: String { message }
}

/// Transparent software fallback: the "sealed" blob is the plaintext, so the
/// persisted key is identical to the portable file-backed default. Used when no
/// hardware element is available (e.g. CI, virtualized hosts). `isHardwareBacked`
/// is false so callers can require hardware and fail closed when configured to.
public struct SoftwareKeyStore: HardwareKeyStore {
    public init() {}
    public var isHardwareBacked: Bool { false }
    public func seal(_ plaintext: Data) throws -> Data { plaintext }
    public func unseal(_ sealed: Data) throws -> Data { sealed }
}

/// Returns the production hardware key store for the current platform: a
/// Secure-Enclave/Keychain-wrapped store on macOS, the software fallback
/// elsewhere.
func makeDefaultHardwareKeyStore(label: String) -> HardwareKeyStore {
    #if canImport(Security) && canImport(CryptoKit) && canImport(Darwin)
    if let store = KeychainHardwareKeyStore(label: label) { return store }
    #endif
    return SoftwareKeyStore()
}

#if canImport(Security) && canImport(CryptoKit) && canImport(Darwin)
/// macOS Secure-Enclave-backed key store.
///
/// A non-exportable P-256 key is created in the Secure Enclave
/// (`kSecAttrTokenIDSecureEnclave`) on first use and persisted in the Keychain
/// (`kSecAttrAccessibleWhenUnlockedThisDeviceOnly`, this-device-only so the
/// secret never syncs or escapes the device). Sealing ECIES-encrypts the proof
/// key to that key's public half; unsealing performs the decryption **inside the
/// Enclave** — the private key material never enters app memory. On hardware
/// without a Secure Enclave the initializer returns nil and the caller falls
/// back to the software store.
///
/// Uses only public Security-framework APIs (App-Store / hardened-runtime safe).
final class KeychainHardwareKeyStore: HardwareKeyStore {

    private let tag: Data
    private let algorithm: SecKeyAlgorithm = .eciesEncryptionCofactorVariableIVX963SHA256AESGCM

    /// Returns nil when the Secure Enclave is unavailable (the caller then uses
    /// the software fallback rather than degrading silently).
    init?(label: String) {
        guard SecureEnclave.isAvailable else { return nil }
        self.tag = Data("io.kseal.proofkey.\(label)".utf8)
    }

    var isHardwareBacked: Bool { true }

    func seal(_ plaintext: Data) throws -> Data {
        let privateKey = try loadOrCreateKey()
        guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
            throw HardwareKeyStoreError(message: "secure-enclave key has no public key")
        }
        guard SecKeyIsAlgorithmSupported(publicKey, .encrypt, algorithm) else {
            throw HardwareKeyStoreError(message: "secure-enclave key does not support the seal algorithm")
        }
        var error: Unmanaged<CFError>?
        guard let cipher = SecKeyCreateEncryptedData(
            publicKey, algorithm, plaintext as CFData, &error
        ) as Data? else {
            throw HardwareKeyStoreError(message: "seal failed: \(Self.describe(error))")
        }
        return cipher
    }

    func unseal(_ sealed: Data) throws -> Data {
        let privateKey = try loadOrCreateKey()
        guard SecKeyIsAlgorithmSupported(privateKey, .decrypt, algorithm) else {
            throw HardwareKeyStoreError(message: "secure-enclave key does not support the unseal algorithm")
        }
        var error: Unmanaged<CFError>?
        guard let plain = SecKeyCreateDecryptedData(
            privateKey, algorithm, sealed as CFData, &error
        ) as Data? else {
            throw HardwareKeyStoreError(message: "unseal failed: \(Self.describe(error))")
        }
        return plain
    }

    // MARK: - Key material

    private func loadOrCreateKey() throws -> SecKey {
        if let existing = try loadKey() { return existing }
        return try createKey()
    }

    private func loadKey() throws -> SecKey? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: tag,
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecReturnRef as String: true,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        switch status {
        case errSecSuccess:
            guard let ref = item else { return nil }
            // Safe: the query requested a SecKey ref.
            return (ref as! SecKey)
        case errSecItemNotFound:
            return nil
        default:
            throw HardwareKeyStoreError(message: "keychain lookup failed: status=\(status)")
        }
    }

    private func createKey() throws -> SecKey {
        var acError: Unmanaged<CFError>?
        guard let access = SecAccessControlCreateWithFlags(
            kCFAllocatorDefault,
            kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            .privateKeyUsage,
            &acError
        ) else {
            throw HardwareKeyStoreError(message: "access control failed: \(Self.describe(acError))")
        }

        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeySizeInBits as String: 256,
            kSecAttrTokenID as String: kSecAttrTokenIDSecureEnclave,
            kSecPrivateKeyAttrs as String: [
                kSecAttrIsPermanent as String: true,
                kSecAttrApplicationTag as String: tag,
                kSecAttrAccessControl as String: access,
            ],
        ]
        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
            throw HardwareKeyStoreError(message: "secure-enclave key creation failed: \(Self.describe(error))")
        }
        return key
    }

    private static func describe(_ error: Unmanaged<CFError>?) -> String {
        guard let error = error?.takeRetainedValue() else { return "unknown error" }
        return CFErrorCopyDescription(error) as String? ?? "unknown error"
    }
}
#endif

File: sdk/desktop/windows/src/Kseal.Desktop/Security/HardwareKeyStore.cs
using System.Runtime.Versioning;
using System.Security.Cryptography;

namespace Kseal.Desktop;

/// <summary>
/// Binds at-rest key material to a hardware-backed key where the platform offers
/// one (Windows TPM via the CNG Platform Crypto Provider), with a clean software
/// fallback. The request-proof HMAC key is sealed before it is persisted, so the
/// on-disk secret cannot be lifted and replayed from another device — yet the
/// <b>request-proof byte layout is unchanged</b>: the core still computes
/// <c>HMAC(proofKey, …)</c>; only how <c>proofKey</c> is protected at rest changes.
///
/// This is the <b>external secure-element mock boundary</b>: tests inject a fake
/// store so the proof-binding logic is exercised deterministically on any host.
/// </summary>
public interface IHardwareKeyStore
{
    /// <summary>
    /// Whether sealed blobs are bound to a hardware-backed key (vs. a software
    /// fallback). Surfaced so a policy can require hardware binding and fail
    /// closed when it is unavailable.
    /// </summary>
    bool IsHardwareBacked { get; }

    /// <summary>Wraps <paramref name="plaintext"/> into an opaque blob safe to persist.</summary>
    /// <exception cref="HardwareKeyStoreException">If the hardware element rejects the operation.</exception>
    byte[] Seal(byte[] plaintext);

    /// <summary>Unwraps a blob previously produced by <see cref="Seal"/>.</summary>
    /// <exception cref="HardwareKeyStoreException">If the blob was not produced by this store/device.</exception>
    byte[] Unseal(byte[] sealedBlob);
}

/// <summary>Raised when a hardware key operation cannot complete.</summary>
public sealed class HardwareKeyStoreException : Exception
{
    public HardwareKeyStoreException(string message, Exception? inner = null) : base(message, inner) { }
}

/// <summary>
/// Transparent software fallback: the "sealed" blob is the plaintext, so the
/// persisted key is byte-identical to the portable file-backed default. Used
/// when no hardware element is available (e.g. CI, virtualized hosts).
/// <see cref="IsHardwareBacked"/> is false so callers can require hardware and
/// fail closed when configured to.
/// </summary>
public sealed class SoftwareKeyStore : IHardwareKeyStore
{
    public bool IsHardwareBacked => false;
    public byte[] Seal(byte[] plaintext) => plaintext;
    public byte[] Unseal(byte[] sealedBlob) => sealedBlob;
}

/// <summary>
/// Returns the production hardware key store for the current platform: a
/// TPM-backed store on Windows when a TPM is present, the software fallback
/// elsewhere (or when no TPM is available).
/// </summary>
internal static class HardwareKeyStoreFactory
{
    public static IHardwareKeyStore Create(string label)
    {
        if (OperatingSystem.IsWindows())
        {
            IHardwareKeyStore? tpm = TpmHardwareKeyStore.TryCreate(label);
            if (tpm is not null) return tpm;
        }
        return new SoftwareKeyStore();
    }
}

/// <summary>
/// Windows TPM-backed key store.
///
/// A non-exportable 2048-bit RSA key is created on first use in the
/// <b>Microsoft Platform Crypto Provider</b> (the TPM-backed CNG provider) and
/// persisted under a tenant/app-scoped key name. Sealing RSA-OAEP-SHA256
/// encrypts the proof key to that key; unsealing performs the decryption
/// <b>inside the TPM</b> — the private key never leaves the chip and the sealed
/// blob is useless on any other machine. When no TPM/provider is available
/// <see cref="TryCreate"/> returns null and the caller uses the software fallback.
/// </summary>
[SupportedOSPlatform("windows")]
public sealed class TpmHardwareKeyStore : IHardwareKeyStore
{
    private const string ProviderName = "Microsoft Platform Crypto Provider";
    private const int KeySizeBits = 2048;

    private readonly string _keyName;

    private TpmHardwareKeyStore(string keyName) => _keyName = keyName;

    /// <summary>
    /// Materializes (or opens) the TPM key, returning null when the TPM/provider
    /// is unavailable so the caller can fall back rather than degrade silently.
    /// </summary>
    public static TpmHardwareKeyStore? TryCreate(string label)
    {
        string keyName = $"io.kseal.proofkey.{label}";
        try
        {
            using CngKey key = OpenOrCreate(keyName);
            return new TpmHardwareKeyStore(keyName);
        }
        catch (CryptographicException) { return null; }
        catch (PlatformNotSupportedException) { return null; }
    }

    public bool IsHardwareBacked => true;

    public byte[] Seal(byte[] plaintext)
    {
        try
        {
            using CngKey key = CngKey.Open(_keyName, new CngProvider(ProviderName));
            using var rsa = new RSACng(key);
            return rsa.Encrypt(plaintext, RSAEncryptionPadding.OaepSHA256);
        }
        catch (CryptographicException e)
        {
            throw new HardwareKeyStoreException("TPM seal failed", e);
        }
    }

    public byte[] Unseal(byte[] sealedBlob)
    {
        try
        {
            using CngKey key = CngKey.Open(_keyName, new CngProvider(ProviderName));
            using var rsa = new RSACng(key);
            return rsa.Decrypt(sealedBlob, RSAEncryptionPadding.OaepSHA256);
        }
        catch (CryptographicException e)
        {
            throw new HardwareKeyStoreException("TPM unseal failed", e);
        }
    }

    private static CngKey OpenOrCreate(string keyName)
    {
        var provider = new CngProvider(ProviderName);
        if (CngKey.Exists(keyName, provider)) return CngKey.Open(keyName, provider);

        var parameters = new CngKeyCreationParameters
        {
            Provider = provider,
            // Persisted machine key; never exportable so the private half is
            // pinned to this TPM.
            KeyCreationOptions = CngKeyCreationOptions.MachineKey,
            ExportPolicy = CngExportPolicies.None,
        };
        parameters.Parameters.Add(new CngProperty(
            "Length", BitConverter.GetBytes(KeySizeBits), CngPropertyOptions.None));
        return CngKey.Create(CngAlgorithm.Rsa, keyName, parameters);
    }
}
